llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp

```diff
@@ -3668,3 +3668,19 @@ bool DwarfDebug::alwaysUseRanges(const DwarfCompileUnit &CU) const {
     return true;
   return false;
 }
+
+void DwarfDebug::beginCodeAlignment(const MachineBasicBlock &MBB) {
+  auto *SP = MBB.getParent()->getFunction().getSubprogram();
+
+  bool NoDebug =
+      !SP || SP->getUnit()->getEmissionKind() == DICompileUnit::NoDebug;
+
+  if (NoDebug)
+    return;
+
+  auto PrevLoc = Asm->OutStreamer->getContext().getCurrentDwarfLoc();
+  Asm->OutStreamer->emitDwarfLocDirective(
+      PrevLoc.getFileNum(), 0, PrevLoc.getColumn(), 0, 0, 0, StringRef());
+  MCDwarfLineEntry::make(Asm->OutStreamer.get(),
```

Comment:
For instructions, usually the line information is recorded by `AsmPrinter::emitFunctionBody() -> DwarfDebug::beginInstruction() -> recordSourceLine() ->  MCStreamer::emitDwarfLocDirective()` and then `MCDwarfLineEntry::make` is called by `MCObjectStreamer::emitInstructionImpl()`.

In the case of nops we aren't going through the normal flow of emitting instructions so we want the line entry for our line information to be emitted immediately otherwise it'll be emitted at the beginning of the next instruction (the end of the nops).

This is also what `MCObjectStreamer::emitInstructionImpl()` does. The line entry is emitted first so that the label points to the beginning of the instruction.

Reply:
OK, so because in this case we're looking at a .align directive rather than a real instruction, it won't be passing through `emitInstructionImpl()` so we need to call `MCDwarfLineEntry::make()` directly. Sounds reasonable.